### ISSUE TITLE
ssh completion can't handle @ in username

### ISSUE THREAD

**Issue by whbdupree** [fish-shell/fish-shell#5959]
Hey folks, 

My work confiscated my linux box and imaged it with their preferred software. They have created a username for me that takes the pattern `workidentity@orgad.org.edu`. Yes that whole string is the username. So when I shell into my machine, i must use the command `ssh workidentity@orgad.org.edu@hostname.dyn.org.edu`. Ugh. I was surprised to see that `ssh` doesn't have a problem with this username. It can parse the whole command, and I can shell into my machine from home. On the other hand, when I try to use the fish ssh autocomplete feature, it recommends me to log into `workidentity@orgad.org.edu`, which is just my username and does not include my hostname and domain. 

I am using fish 3.0.2 on archlinux, and I was able to reproduce it without using any third party plugins. I don't have any control over my username at this point, so changing it isn't an option.

**Comment by faho** [comment]
This... basically isn't handleable?

The problem is that we don't have a user database. We could look up the local users in passwd, but that wouldn't include this one.

So we have no way to know that this is a username (it could be a valid user@host *as well*), so we assume that it's a user@host.

However, what already works is just typing the second @ - we assuming everything up to the *last @* is the user, so we take the rest as the host.

**Comment by whbdupree** [comment]
ssh can handle it just fine

**Comment by faho** [comment]
Yes, it can. But we *can't*, because we don't have a way to look up that it's a valid user.

**Comment by whbdupree** [comment]
The whole username@hostname part of the command is a substring with no spaces. Instead of searching from the start for the first @, you could search from the end of that substring and then assume that everything preceding that space-less substring is a username.

**Comment by faho** [comment]
*we already do*.

The problem is if you have *just that* it looks *exactly* like a "user@host", *AND IT COULD BE A VALID USER@HOST*!

**Comment by faho** [comment]
E.g. your username is "workidentity@orgad.org.edu".

The problem is that "orgad.org.edu" is a hostname, and that host presumably actually exists (and even if it didn't, there'd be no real way to tell other than trying to connect to it, which we don't want to do, and it might still be a correct argument).

So that means that the full "workidentity@orgad.org.edu" is potentially a correct thing to connect with, which means we should then complete other things. The only way to tell that it's not a correct thing to connect with is to be the ssh server on orgad.org.edu, and *it* won't tell you, because you don't allow unauthorized remotes to enumerate users.

In other words: We *cannot* handle this, it is inherently ambiguous.

**Comment by faho** [comment]
Note that you can make your own wrapper function that completes the rest:

```fish
function my-ssh --wraps 'ssh "workidentity@orgad.org.edu"@'
    ssh "workidentity@orgad.org.edu"$argv[1] $argv[2..-1]
end
```

**Comment by zanchey** [comment]
FWIW, I just use `.ssh/config` to set my username for all hosts matching a particular pattern.
